source/app/controllers/users_controller.rb

```diff
@@ -1,4 +1,38 @@
 class UsersController < ApplicationController
+  def index
+    logged_in? ? redirect_to(current_user) : redirect_to(login_path)
+  end
+
   def show
+    if owns_account? or is_admin?
+      @user = User.find(params[:id])
+    else
+      redirect_to login_path
+    end
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      user_id = User.find_by(user_name: @user[:user_name]).id
+      session[:user_id] = user_id
+      redirect_to current_user
+    else
+      redirect_to signup_path
```

Comment:
Consider `render`ing the view at `signup_path` instead so you can show validation errors from `@user`. The redirect loses the context.